### ISSUE TITLE
[Bug]: window.location is not equal to useLocation() with loader in react 17

### ISSUE THREAD

**Issue by holynewbie** [remix-run/react-router#10446]
### What version of React Router are you using?

 6.11.1

### Steps to Reproduce

package.json

```json
{
  "name": "codesandbox-template-vite",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "tsc && vite build",
    "preview": "vite preview"
  },
  "dependencies": {
    "react": "^17.0.0",
    "react-dom": "^17.0.0",
    "react-router-dom": "^6.11.1"
  },
  "devDependencies": {
    "@types/react": "^17.0.0",
    "@types/react-dom": "^17.0.0",
    "@vitejs/plugin-react": "^2.0.0",
    "typescript": "^4.6.4",
    "vite": "^4.3.0"
  }
}
```

main.tsx
```ts
import React from "react";
import ReactDOM from "react-dom";
import {
  createBrowserRouter,
  RouterProvider,
  useNavigate,
  Outlet,
  useLocation,
} from "react-router-dom";

const App = () => {
  const nav = useNavigate();
  return (
    <div>
      <button onClick={() => nav("/abc?query=123")}>跳转</button>
      <Outlet />
    </div>
  );
};
const Abc = () => {
  const location = useLocation();
  console.log(
    "window.location: {pathname: %s, search: %s}",
    window.location.pathname,
    window.location.search
  );
  console.log(
    "useLocation(): {pathname: %s, search: %s}",
    location.pathname,
    location.search
  );

  return <div>abc</div>;
};

const router = createBrowserRouter([
  {
    path: "/",
    element: <App />,
    children: [
      {
        path: "abc",
        element: <Abc />,
        loader() {
          // whatever code here
          return null;
        },
      },
    ],
  },
]);

ReactDOM.render(
  <React.StrictMode>
    <RouterProvider router={router} />
  </React.StrictMode>,
  document.getElementById("root") as HTMLElement
);
```

the result is wrong，`window.location` get older state。 But if I  annotate the `loader`， it turns right.

### Expected Behavior

`window.location` has same `pathname`、`search` ... with `useLocation`

### Actual Behavior

![screenshot-20230504-145110](https://user-images.githubusercontent.com/8347286/236131460-dcf8e972-b4ea-4a5a-9081-2b905e272efc.png)

https://user-images.githubusercontent.com/8347286/236131487-73145145-3c34-4ef8-a1d2-d41af65c6746.mov

**Comment by sash9696** [comment]
Hi @brophdawg11 , I would like to work on this issue. It would be of great help if you could assign it .

**Comment by timdorr** [comment]
Issues don't get assigned on public projects unless only one person can solve it. @holynewbie already submitted a PR.

**Comment by brophdawg11** [comment]
I think we can go ahead and merge the linked PR, but want to document here just for clarity that apps should **not** be relying on `window.location` and should always reference `useLocation` when possible.  No matter what we do `window.location` will always be out of sync at one point or another (`popstate` events, concurrent mode, etc.) so this will fix this specific React 17 issue - but it will _not_ make window.location a reliable source of data in a React Router app :)

**Comment by brophdawg11** [comment]
This is fixed by #10448 but didn't quite get merged in time for the pending `6.12.0` release, so it will be available on the next release after that goes out - likely  `6.12.1` or `6.13.0`
